### PR TITLE
planner: horror-hack: create fake cylinder at end of cylinder table

### PR DESCRIPTION
**Pull request by bstoeger**
When we had fixed-sized cylinder arrays, the planner used the last
empty cylinder for "surface air". This was not recognized by the UI
as a separate cylinder, because "empty cylinder" was the sentinel for
the end of the table. The conversion to dynamically sized cylinder
tables broke this code: everytime the surface segment is changed,
a new dummy cylinder is added, which is visible in the UI.

As a very temporary stop-gap fix, emulate the old code by creating
a cylinder and then setting the end-of-table to before that cylinder.
This means that we have to loosen the out-of-bound checks.

That's all very scary and should be removed as soon as possible.

Signed-off-by: Berthold Stoeger <bstoeger@mail.tuwien.ac.at>

### Describe the pull request:

- [x] Bug fix
- [ ] Functional change
- [ ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This is supposed to sketch a temporary fix to #2788. I feel very bad for this and would prefer a proper fix. However, that might be too short before release.

### Changes made:

Create dummy cylinder at end of cylinder table for "surface air".

### Related issues:

see discussion in #2788

### Mentions:

@atdotde

### CODE REVIEW

**Comment by atdotde** [comment]
I don't think this is the correct approach. I think the problem is in handling the cylinders back and forth between dive and dive plan.

Please give me some more time to debug this.

**Comment by bstoeger** [comment]
> I don't think this is the correct approach.

I am sure that it is not. However, it emulates the old code that did kind of work.

> Please give me some more time to debug this.

Sure - take your time.

**Review by atdotde — COMMENTED** [review]
After a night of sleep, I came to terms with this approach, maybe it's not so bad after all. But I would do the semantics slightly differently: Let's start from the statement "there is always air at the surface and it comes in an infinite amount". So it makes some sense to have it as a default entry in a cylinder table that is never explicitly shown. I would make the table always contain element [cylinders.nr], i.e. one bigger but that entry understood as read only filled with only 0s. This could then as well be used for surface segments. But I would make it always there, not to be initialised from the planner. You can just refer to it explicitly by making cylinder id = cylinders.nr. 
Yes there is a cost in RAM, reserving space for something that is always 0, but I think it's worth it, so if you have to read form that cylinder, there is something there.

I would like to play with this a little more.

File: core/planner.c [hunk 1]
```diff
@@ -272,7 +272,9 @@ static void create_dive_from_plan(struct diveplan *diveplan, struct dive *dive,
 		if (dp->cylinderid != lastcylid) {
 			/* need to insert a first sample for the new gas */
 			add_gas_switch_event(dive, dc, lasttime + 1, dp->cylinderid);
-			cyl = get_or_create_cylinder(dive, dp->cylinderid);
+			cyl = get_cylinder(dive, dp->cylinderid); // FIXME: This actually may get one past the last cylinder for "surface air".
+			if (!cyl)
+				continue;
```
Comment (atdotde) [inline core/planner.c:278]: This should not happen. Silently going to the next dp is not the right thing. There should at least be a warning.

File: core/planner.c [hunk 2]
```diff
@@ -1072,10 +1074,17 @@ bool plan(struct deco_state *ds, struct diveplan *diveplan, struct dive *dive, i
...
+		// we had a fixed cylinder table: It makes sure that there is at least
+		// one cylinder beyond the used cylinders in the cylinder table.
+		// It then lets the cylinder point past the end of the table. Thus,
+		// this fake-cylinder can be used by the planner, but is not visible to
+		// the UI. Fix this as soon as possible!
 		cylinder_t cyl = empty_cylinder;
 		cyl.cylinder_use = NOT_USED;
 		add_to_cylinder_table(&dive->cylinders, dive->cylinders.nr, cyl);
 		current_cylinder = dive->cylinders.nr - 1;
+		dive->cylinders.nr--;
```
Comment (atdotde) [inline core/planner.c]: As I said, I would vote for always having an extra 0 cylinder at the end, so we are not decreasing the number in every run but simply use that last cylinder.

**Comment by bstoeger** [comment]
I still think this approach is utterly broken. Since we are under time-pressure I can indeed simply always have the cylinder table have one more entry (memory usage is not an issue at all - we used to have 20 cylinders for *every* dive and nobody ever complained). I will push something to that effect in the later afternoon. However, this opens a can of worms:

What happens when the dive is saved and loaded? Are there gas change events to the one-past-last cylinder? What will the git and xml loaders do in such a case? What will fixup_dive() do? We have to be extremely careful not to end up with broken dives on disk.

Ultimately, I really think there should be the possibility to represent gas switch to surface air. In a quick test this seemed to kind-of work. However, there were a few problems. Notably in some parts of the code cylinderid = -1 is interpreted as "use the same cylinder". So either we need a flag or a different magic number. In any case, certainly not for this release.

**Comment by bstoeger** [comment]
Since I found a few seconds, here a short update where there is always an empty cylinder at the end of the array.

As I feared, saving does not work properly, at least to XML:
- Plan a dive with surface segment
- Save, close application load
- Replan dive

The surface segment is now missing, probably because the invalid cylinder id confuses the code.

**Comment by bstoeger** [comment]
> The surface segment is now missing, probably because the invalid cylinder id confuses the code.

Ok, so more testing gives a strange behavior: When replanning a dive, the surface interval is set to the last used value, not the value of the replanned dive.

So for example:
1) plan dive, set surface interval to 10 min
2) plan dive, set surface interval to 20 min
3) replan first dive, surface interval is now 20 min, because that was the last value used in the planner.

Is that intended behavior?

**Comment by dirkhh** [comment]
structurally I really struggle with the idea that we special case a cylinder without having a flag that clearly says "THIS IS NOT A CYLINDER". Which is different from "NOT USED". Having a magic last cylinder will require changes in a ton of places - at least when we make these changes, let's give ourselves a clear way of knowing that this is indeed just a magic cylinder that doesn't actually exist. I believe that doing that would then also allow us to not force this as the last cylinder which would likely make things easier elsewhere?
But that of course means that the model knows about the cylinder and the UI layer and the storage layer know not to show / save the data.

I still wonder if it wouldn't be easier to simply use '-1' (or heck, '-42') to indicate 'surface air' and to instead special case THAT in the deco calculations. That sounds like a much more self contained set of changes.

**Comment by bstoeger** [comment]
> structurally I really struggle with the idea that we special case a cylinder without having a flag that clearly says "THIS IS NOT A CYLINDER". Which is different from "NOT USED". Having a magic last cylinder will require changes in a ton of places - at least when we make these changes, let's give ourselves a clear way of knowing that this is indeed just a magic cylinder that doesn't actually exist. I believe that doing that would then also allow us to not force this as the last cylinder which would likely make things easier elsewhere?

From your message, I'm not 100% sure if you are aware of what this PR does: The point is that this cylinder is _not seen_ by the rest of the code, because it is beyond `cylinders.nr`. This is what the old code _effectively_ did, so this here is probably be the least invasive version that is currently possible. Since the cylinder is not seen by the rest of the system, there is little point in arguing over how we should mark it - it purely virtual. That's why this PR has to weaken some bound-checks.

> I still wonder if it wouldn't be easier to simply use '-1' (or heck, '-42') to indicate 'surface air' and to instead special case THAT in the deco calculations. That sounds like a much more self contained set of changes.

Yes, that's what I'm saying all the time. However, that will be distinctly more invasive than this PR. I have a proof of concept that seems to kind-of work. It will need some auditing though, because some parts of the code treat '< 0' as use the same cylinder, so this might give some skewed results (e.g. wrong cylinder use, etc.)

**Comment by dirkhh** [comment]
> From your message, I'm not 100% sure if you are aware of what this PR does: The point is that this cylinder is _not seen_ by the rest of the code, because it is beyond `cylinders.nr`. This is what the old code _effectively_ did, so this here is probably be the least invasive version that is currently possible. Since the cylinder is not seen by the rest of the system, there is little point in arguing over how we should mark it - it purely virtual. That's why this PR has to weaken some bound-checks.

You are correct. I did not realize that that's what the PR does. Need to re-read.

**Comment by bstoeger** [comment]
> I still wonder if it wouldn't be easier to simply use '-1' (or heck, '-42') to indicate 'surface air' and to instead special case THAT in the deco calculations. That sounds like a much more self contained set of changes.

One problem with that is that we can't use normal gas-change events, because `validate_gaschange()`, which is called during `fixup_dive()` will try to "correct" these event.

One thing we could do is misuse the `SAMPLE_EVENT_SURFACE` event to mean switch to surface air in planner context. The problem is that apparently some dive computers emit these events when the diver briefly breaches the surface and we don't want to interpret these as "stopped using gas".

Or use gas index "-1" as "unknown" and "-2" as "no gas". These will end up in the on-disk logs as "0" and "-1", respectively.

**Comment by bstoeger** [comment]
@atdotde, @dirkhh: If Dirk wants to do a release this week, we will not be able to properly fix this and we need a stop-gap measure. I believe that this PR is the least intrusive that we can achieve in a short time. Still we need to do some auditing, concerning e.g. what is written/loaded to/from XML and git.

I can tone down the comments and commit messages if you prefer. I can also try to summarize (my view of) the problem and different ideas that I have how we can fix it in the long run. However for new, we need a decision of how to proceed in the short term.

**Comment by dirkhh** [comment]
I think this is a reasonable approach. I'm not too worried about the tone anywhere. I'm worried about getting this tested and making sure we don't store nonsense to either git or XML that later bites us. That should be easy to test, I hope. Add some plans with surface intervals, mess around with a few things, store, inspect what was stored.

**Comment by dirkhh** [comment]
I'll ask you to merge this yourself when you think it's ready

**Comment by atdotde** [comment]
I am currently looking at this.

**Comment by atdotde** [comment]
To me, this seems to work. I propose merging for the time being.

**Comment by bstoeger** [comment]
I tried this and two points of note:

1) When saving I get the following events in the XML file: `<event time='21:43 min' type='11' value='21' name='gaschange' cylinder='0' />` at the beginning of the surface segment. Which is arguably wrong, but I'm not sure if that is different from what the code did before.

2) I can easily mess up the profile when replanning a dive. For some reason, I can then switch between dives and the events, but not the profile data is updated!?

**Comment by bstoeger** [comment]
Arghl - never mind. I was testing the wrong branch (my event-changes).

**Comment by bstoeger** [comment]
Ok - next test, this time the correct branch:

- We now indeed write cylinder change events to a non-existing cylinder (one past last). That still raises the question: is this what we did before? (Ubuntu ships with a broken version!)

- The profile is still easy to mess up by replanning a dive:
screenshot

**Comment by bstoeger** [comment]
You don't even need to replan a dive for that bug. A long surface segment is sufficient to trigger it. See attached log: switch between both dives.
[test2.zip](https://github.com/subsurface/subsurface/files/4560548/test2.zip)

**Comment by dirkhh** [comment]
so... is this bug in 4.9.3?
I'm trying to get 4.9.4 out. I'd prefer if we didn't regress, but we also can't make "every bug we find is fixed" a precondition...
grabbing a 4.9.3 build and will test...

**Comment by dirkhh** [comment]
Yep, same broken profile in 4.9.3 on my Mac.

So let's file this as a bug and move on...

**Comment by bstoeger** [comment]
Since you have a 4.9.3 at hand (mine doesn't run since it was linked against the wrong libgit2), could you plan a dive with surface long surface segment and look what gas-change event you get in the XML?

Thanks.

**Comment by atdotde** [comment]
Maybe I am stating the obvious, but what seems to happen with the too long surface interval is that part of another dive is plotted.

**Comment by dirkhh** [comment]
> Since you have a 4.9.3 at hand (mine doesn't run since it was linked against the wrong libgit2), 

I shouldn't laugh... but this is exactly why I like Mac/Windows apps and AppImages. I can go back and run ancient versions of Subsurface with those :-)

> could you plan a dive with surface long surface segment and look what gas-change event you get in the XML?

I get none at all, but maybe I did that wrong...
```
<dive number='3' date='2020-04-30' time='23:40:07' duration='15:53 min'>
[deleted notes]
  <cylinder size='11.094 l' workpressure='206.843 bar' description='AL80' start='206.843 bar' end='126.805 bar' depth='65.73 m' />
  <divecomputer model='planned dive' last-manual-time='20:00 min'>
  <depth max='20.0 m' mean='13.453 m' />
  <surface pressure='1.013 bar' />
  <sample time='0:00 min' depth='0.0 m' pressure='206.843 bar' />
  <sample time='0:52 min' depth='20.0 m' pressure='203.598 bar' />
  <sample time='1:47 min' depth='20.0 m' pressure='198.46 bar' />
  <sample time='6:51 min' depth='20.0 m' pressure='170.147 bar' />
  <sample time='8:55 min' depth='0.0 m' pressure='162.562 bar' />
  <sample time='15:46 min' depth='0.0 m' pressure='150.018 bar' />
  <sample time='17:37 min' depth='15.0 m' pressure='144.128 bar' />
  <sample time='20:00 min' depth='15.0 m' pressure='133.317 bar' />
  <sample time='21:08 min' depth='4.8 m' pressure='129.853 bar' />
  <sample time='22:44 min' depth='0.0 m' pressure='126.805 bar' />
  </divecomputer>
</dive>
```

**Comment by bstoeger** [comment]
> Maybe I am stating the obvious, but what seems to happen with the too long surface interval is that part of another dive is plotted.

Yes, if you resize the window, the correct data is plotted. Strange.

**Comment by bstoeger** [comment]
And when I remove the "new" gas-change event with one-past-last cylinder, the display bug goes away. Therefore, it *is* a regression introduced with this PR. :(

**Comment by bstoeger** [comment]
> I get none at all, but maybe I did that wrong...

Then what I don't understand: if there is no gas-change event, how does the code know that we are at surface air? Does it special-case 0 depth? And if that is the case: why do we have to switch to a virtual cylinder in the first place? I don't get it... Robert?

**Comment by bstoeger** [comment]
Ok, I see - if we don't change to a fake cylinder it keeps consuming gas. Still don't understand how the old code worked. Sigh.

**Comment by bstoeger** [comment]
@dirkhh: The 4.9.3 appimage is already broken with respect to surface intervals. Where can I get older versions?

**Comment by dirkhh** [comment]
> @dirkhh: The 4.9.3 appimage is already broken with respect to surface intervals. Where can I get older versions?

https://subsurface-divelog.org/downloads/
all the way back to 4.5.2 :-)

**Comment by bstoeger** [comment]
Yep - even 4.9.1 is broken:
1) Started with an empty log.
2) Plan dive - add surface interval. Looks good.
3) Accept plan - got the strange profile from the attached screen shot.

4) There is a spurious cylinder in the equipment tab.
5) Save log. There is indeed a second cylinder with "unused" in the log.
6) Restart application and load log. The second cylinder is gone. Profile looks good.
7) Save log. The second cylinder is now gone from the XML file. The funny thing though: the profile looks fine now. No more air consumption in the surface segment...

Thus the question remains: how can we achieve 7) directly without doing cylinder tricks..?

In conclusion, this probably never really worked properly. (Note: this is not supposed to be an excuse for the mess I produced when removing MAX_CYLINDERS - I totally misread the code.)

**Comment by dirkhh** [comment]
does that mean that we should merge this and consider ourselves ready for 4.9.4?
And then fix the longstanding issue over the next few weeks?

**Comment by bstoeger** [comment]
Not yet please - this is better than master but worse than 4.9.3, since it produces 1) rendering errors on the profile and 2) kind-of invalid XML files with cylinder-ids past the last cylinder.

**Comment by dirkhh** [comment]
thanks for that clarification

**Comment by bstoeger** [comment]
> this is better than master but worse than 4.9.3

Well, I was wrong. This appears to produce *exactly* the behavior of earlier versions. With 4.9.1 and the steps above I get the XML file:
```
<dive number='1' date='2020-05-01' time='00:02:55' duration='21:54 min'>
  [...]
  <cylinder size='11.1 l' workpressure='207.0 bar' description='unknown' start='207.0 bar' end='109.727 bar' depth='66.0 m' />
  <divecomputer model='planned dive' last-manual-time='20:00 min'>
  <depth max='15.0 m' mean='14.064 m' />
  <surface pressure='1.013 bar' />
  <event time='21:55 min' type='11' value='21' name='gaschange' cylinder='1' />
  <sample time='0:00 min' depth='0.0 m' pressure='207.0 bar' />
  <sample time='1:00 min' depth='15.0 m' pressure='203.723 bar' />
  <sample time='20:00 min' depth='15.0 m' pressure='114.996 bar' />
  <sample time='20:42 min' depth='10.8 m' pressure='112.549 bar' />
  <sample time='21:54 min' depth='0.0 m' pressure='109.727 bar' />
  <sample time='21:55 min' depth='0.0 m' />
  <sample time='43:54 min' depth='0.0 m' />
  </divecomputer>
</dive>
```
Note that there is an gaschange-event to cylinder 1, which doesn't exist. And if you switch between dives, the profile is broken:


This is therefore probably indeed the right thing to do for now, as it is not more broken than earlier releases (and not too many people complained). Nevertheless, please give me some more time to think about this. At least the profile bug should be fixed...

**Comment by bstoeger** [comment]
Oh please - debugging this is *not* fun.

When I remove the following lines from `calculate_max_limits_new()` in `profile.c`:
```
                /* Make sure we can fit all events */
                ev = dc->events;
                while (ev) {
                        if (ev->time.seconds > maxtime)
                                maxtime = ev->time.seconds;
                        ev = ev->next;
                }
```
This decreases the `maxtime` variable by 1(!) second. This single second has the effect that the profile is properly updated. In return, the surface segment is not shown anymore. Indeed, a few lines up we read:
```
                        if ((depth > SURFACE_THRESHOLD || lastdepth > SURFACE_THRESHOLD || in_planner()) &&
                            s->time.seconds > maxtime)
                                maxtime = s->time.seconds;
```
Thus, apparently surface segments should not be shown?

Thus, my current theory is that the profile gets confused by the fact that `maxtime` is one second inside the last segment, but that the segment extends significantly farther...?

**Comment by atdotde** [comment]
Indeed, surface segments (defined w.rt. SURFACE_THRESHOLD) should not be show for logged dives (apparently, some dc's record those at the end of the dive before turning off). Or put differently, (in my understanding), it's ok if those extend out of the display. This is different in the planner and you could argue about planned dives in the log. But even for those, outside the planner, you see that the heat-map is not proper for the surface segment (it doesn't change with time during the segment as there, we don't do interpolation). I always considered that not worrisome as it wasn't really visible in the log (as opposed to the planner).

**Comment by atdotde** [comment]
What I am currently looking at is, I inserted a
 maxime /= 2;
after the code segment you quoted that increases the maxtime to see if it has any effect besides not making the dive fit the profile widget. It does (it doesn't when you *=2), for example we force going to the surface at maxim... and the artefacts look pretty much like what we are discussing here.

**Comment by atdotde** [comment]
Have you noticed that also the information in the inbox when you move the mouse is corrupted (look for example at the time)?

**Comment by bstoeger** [comment]
> Indeed, surface segments (defined w.rt. SURFACE_THRESHOLD) should not be show for logged dives (apparently, some dc's record those at the end of the dive before turning off). Or put differently, (in my understanding), it's ok if those extend out of the display. 

Understood - thanks for the information.

> Have you noticed that also the information in the inbox when you move the mouse is corrupted (look for example at the time)?

Indeed, a segment that starts before but extends beyond `maxtime` seems to somehow confuse the profile.

**Comment by atdotde** [comment]
I am currently looking at populate_plot_entries in profile.c . I suspect that is not happy when you do weird stuff to maxtime (is in: it stops computing plot entries).

**Comment by bstoeger** [comment]
Yes, it appears like it inherently can't handle the last segment extending much beyond `maxtime`. When allocating the points:
```
        nr = dc->samples + 6 + maxtime / 10 + count_events(dc);
```
It may not even allocate enough points for the last segment, because it is based on maxtime. Thus, I think the calculation of `maxtime` should be changed to include the last sample after the last event...?

**Comment by bstoeger** [comment]
I added a commit, which tries to fix the `maxtime` calculation. Apparently, `maxtime` should be rounded to a segment boundary. This seems to work in a first quick test.

**Comment by atdotde** [comment]
Yes, this looks good. As a bonus, it also gets the heat map for the surface segment right.

**Comment by bstoeger** [comment]
Few, that's a relief - thanks for testing. Hopefully this doesn't break other use cases...!

**Comment by atdotde** [comment]
By the way, I now keep getting the can't find a spot in the dataModel warning. Not exactly sure what causes that.

**Comment by bstoeger** [comment]
Surely a fallout of #2800 - another thing to checkout. Help.

**Comment by bstoeger** [comment]
@dirkhh wanted us to commit this thing when ready, so let's do it.

**Comment by atdotde** [comment]
It seems it happens during animations when flipping between dives. Looking into this but I am not too worried.

**Comment by atdotde** [comment]
Fingers crossed

**Comment by bstoeger** [comment]
> It seems it happens during animations when flipping between dives. Looking into this but I am not too worried.

Interesting. I'm quite sure that always happened, we just haven't seen the message so far.

**Comment by atdotde** [comment]
I agree.

**Comment by atdotde** [comment]
My suspicion is that while animating, we try to figure out the position for events that are no longer there: There is still the event marker (with its time) that is trying figure out where to go but the underlying dive already changed so it cannot determine the corresponding sample.
